Add fragmentation propagation test

claytron5000 opened the pull request:
#7958
I'm not convinced I totally understand the specification on this one.

> In a column flex container, the break-before property on the first item and the break-after property on the last item are propagated to the flex container. Forced breaks on other items are applied to the item itself. 

This fails on Chrome Canary, and I haven't seen it work in Firefox either.

<!-- Reviewable:start -->

<!-- Reviewable:end -->

claytron5000 committed:
Add fragmentation propagation test

fantasai reviewed (changes requested):
OK, so this is totally wrong. :) The spec isn't saying that the computed values are passed up; it's talking about a different, abstract mechanism which is unrelated to computed values. What it's doing is making sure that a forced break-before or break-after doesn't create a break between a parent and its first/last child--breaks are only supposed to only happen between siblings, otherwise it looks weird with dangling empty bits of broken borders/padding. See https://www.w3.org/TR/css-break-3/#forced-breaks

To test this, you actually have to test layout using the reftest infrastructure. IIRC WPT doesn't have pagination reftest infrastructure, though (even though Mozilla has it), so the best you can do atm is write a test that'll check fragmentation in a multicol environment. It won't be able to test combined break types, but it can check for basic propagation. :/

(Might be worth trying to make the spec wording a bit clearer it seems, so I filed https://github.com/w3c/csswg-drafts/issues/2614 )

claytron5000 commented:
@fantasai Thanks for the clarification, I had a feeling I was off base there.

If I'm reading this right, are you saying we're checking to make sure the first child element is not breaking to the second column(for example), or broken padding/border etc?

claytron5000 committed:
Refactor test

claytron5000 commented:
In this version I'm testing for collapsing borders/padding on the top and left side of the container. I have to think about how to test the `break-after` situation.

gregwhitworth commented:
@fantasai thanks for jumping in and helping out here. Multicol was the only way I could think of to potentially test this. Here is an example of this, and this is why writing the testcases is so important for interop: http://jsbin.com/mewotuzugi/edit?html,css,output

![image](https://user-images.githubusercontent.com/865244/39322387-93f5a38e-493e-11e8-8145-5319fc56d98f.png)

Reading the spec text, Chrome/Edge are doing what I would expect them to do in the regular column fragmentation scenario. Edge supports avoiding column breaks, which makes testing the above a bit easier as you'll want the first item to break on the column: http://output.jsbin.com/feyibulefe/1

![image](https://user-images.githubusercontent.com/865244/39322556-1ac57fba-493f-11e8-8d79-747a464fba34.png)

Based on the initial auto values and the way other fragmentation is occurring across pages, etc, I believe Edge is doing the intended behavior - please correct me if I'm wrong here @fantasai.

If this is indeed the case, then you could build a reftest that mimics this but you could also try using `getBoundingClientRects` to ensure its position and size.

fantasai commented:
@gregwhitworth As I mentioned on IRC, you cannot test this case by having the flex container be the first child of the multicol: a forced break there will have no effect because it's already at the top of the first column. You need to put content before the flex container to see if a break is taken between that content and the flex container.

The goal of this clause in the spec is to prevent any breaks between the start of the flex container and the first item, even if the first item requests a forced break before: that forced break gets propagated to the flex container, moving the flex item to the top of the next column by pushing the entire flex container over.

Edge is definitely getting it wrong (I'm assuming it's the black UA)--it's putting a break between the top content edge of the flex container and the top margin of the flex item, and that's not allowed.

gregwhitworth commented:
@fantasai thank you so much for the clarification.

@claytron5000 Based on the above, do you feel like you have enough info to give a test a go? If not - I can tackle it when I get a chance. Let me know.

claytron5000 commented:
> The goal of this clause in the spec is to prevent any breaks between the start of the flex container and the first item, even if the first item [*in the flex container] requests a forced break before: that forced break gets propagated to the flex container, moving the flex item to the top of the next column by pushing the entire flex container over.

I'm starting to understand what's going on. I added one clarifying phrase above in brackets.

In this example I've added a "firsty" element that in Firefox bumps the entire container into the second column (unexpected/unwanted behavior), but in Chrome, does not bump it over. (I don't have Edge on this computer).
https://codepen.io/claytron5000/pen/BxQddb?editors=1100

FireFox:
<img width="1438" alt="image" src="https://user-images.githubusercontent.com/5767672/39365735-071a8d18-4a00-11e8-8c03-888ac6a13388.png">
Chrome:
<img width="1437" alt="image" src="https://user-images.githubusercontent.com/5767672/39365747-107c4158-4a00-11e8-9d1a-7645cc6129c6.png">

Note: There seems to be a secondary behavior irregularity illustrated in @gregwhitworth 's bins that is independent(though maybe technically related) of this issue where Firefox doesn't push the `break-after` items to the second column.

gregwhitworth commented:
I don't think that we can trust what Firefox is doing here as they don't seem to ever actually fragment a column direction flex container (but I do agree with your premise). Here is a side-by-side of Chrome & Edge in your test (make sure to remove `break-before: column` as Edge supports this and will incorrectly "pass" your test, but in this screenshot it is removed):

![image](https://user-images.githubusercontent.com/865244/39370111-f0410818-49f2-11e8-80af-419455d74e10.png)

Based on my reading of the spec and @fantasai clarification, I think that Firefox's end result is what is expected, but based on the general case I think they're doing that solely because the entirety of the flex container doesn't fit so they move it over. So I think it should look like this (according to the spec - this is using block layout [not flex] and can be the basis for the ref tests):

![image](https://user-images.githubusercontent.com/865244/39370764-9ae2c2d8-49f4-11e8-9c1a-a2764c01d840.png)

Here is the codepen of that: https://codepen.io/gregwhitworth/pen/MGbrpa

fantasai commented:
Also, and this is totally not required if you don't wanna, but, can we use somewhat less garish colors for the testcase? :) I tend to use orange/blue/aqua for things that are important to pay attention to in the testcase and silver/gray/black for incidentals. ^_^
